File: rust/src/world.rs
use std::collections::HashMap;
use rand::Rng;

pub struct World<'a> {
  width: u32,
  height: u32,
  pub tick: u32,
  cells: HashMap<String, Cell<'a>>,
  cached_directions: [[i8; 2]; 8],
}

impl<'a> World<'a> {

  pub fn new(width: u32, height: u32) -> World<'a> {
    let mut world = World {
      width,
      height,
      tick: 0,
      cells: HashMap::new(),
      cached_directions: [
        [-1, 1],  [0, 1],  [1, 1], // above
        [-1, 0],           [1, 0], // sides
        [-1, -1], [0, -1], [1, -1] // below
      ],
    };

    world.populate_cells();
    world.prepopulate_neighbours();

    world
  }

  pub fn _tick(&'a mut self) {
    // First determine the action for all cells
    for (_key,mut cell) in self.cells {
      let alive_neighbours = self.alive_neighbours_around(&mut cell);
      if !cell.alive && alive_neighbours == 3 {
        cell.next_state = Some(1);
      } else if alive_neighbours < 2 || alive_neighbours > 3 {
        cell.next_state = Some(0);
      }
    }

    // Then execute the determined action for all cells
    for (_key,mut cell) in self.cells {
      if cell.next_state == Some(1) {
        cell.alive = true;
      } else if cell.next_state == Some(0) {
        cell.alive = false;
      }
    }

    self.tick += 1;
  }

  // Implement first using string concatenation. Then implement any
  // special string builders, and use whatever runs the fastest
  pub fn render(&'a self) -> String {
    let mut rendering = String::new();
    for y in 0..self.height {
      for x in 0..self.width {
        // unwrap pulls the Cell out of an Option<>
        let cell = self.cell_at(x, y).unwrap();
        rendering += &cell.to_char().to_string();
      }
      rendering += "\n";
    }
    rendering
  }

  fn populate_cells(&'a mut self) {
    for y in 0..self.height {
      for x in 0..self.width {
        let alive = rand::thread_rng().gen::<f32>() <= 0.2;
        self.add_cell(x, y, alive);
      }
    }
  }

  fn prepopulate_neighbours(&'a mut self) {
    for (_key,cell) in self.cells {
      self.neighbours_around(&mut cell);
    }
  }

  fn add_cell(&'a mut self, x: u32, y: u32, alive: bool) -> &Cell<'a> {
    // TODO: Custom exceptions

    let cell = Cell::new(x, y, alive);
    let key = String::from(format!("{}-{}", x, y));
    self.cells.insert(key, cell);
    // unwrap pulls the Cell out of an Option<>
    &self.cell_at(x, y).unwrap()
  }

  fn cell_at(&'a self, x: u32, y: u32) -> Option<&Cell<'a>> {
    let key = String::from(format!("{}-{}", x, y));
    self.cells.get(&key)
  }

  fn neighbours_around(&'a self, cell: &'a mut Cell<'a>) -> &'a Vec<&'a Cell<'a>> {
    if cell.neighbours.is_none() { // Must return a boolean
      let mut neighbours: Vec<&Cell<'a>> = Vec::new();
      for set in self.cached_directions.iter() {
        let neighbour = self.cell_at(
          cell.x + set[0] as u32,
          cell.y + set[1] as u32,
        );

        if neighbour.is_some() {
          // unwrap pulls the Cell out of an Option<>
          neighbours.push(neighbour.unwrap());
        }
      }
      cell.neighbours = Some(neighbours);
    }

    // unwrap pulls the Cell out of an Option<>
    cell.neighbours.as_ref().unwrap()
  }

  // Implement first using filter/lambda if available. Then implement
  // foreach and for. Retain whatever implementation runs the fastest
  fn alive_neighbours_around(&'a self, cell: &'a mut Cell<'a>) -> u8 {
    let mut alive_neighbours = 0 as u8;
    let neighbours = self.neighbours_around(cell);
    for i in 0..neighbours.len() {
      let neighbour = neighbours[i];
      if neighbour.alive {
        alive_neighbours += 1;
      }
    }
    alive_neighbours
  }

}

struct Cell<'a> {
  x: u32,
  y: u32,
  alive: bool,
  next_state: Option<u8>,
  neighbours: Option<Vec<&'a Cell<'a>>>,
}

impl<'a> Cell<'a> {

  pub fn new(x: u32, y: u32, alive: bool) -> Cell<'a> {
    let cell = Cell {
      x,
      y,
      alive,
      next_state: None,
      neighbours: None,
    };

    cell
  }

  pub fn to_char(&'a self) -> char {
    if self.alive { 'o' } else { ' ' }
  }

}

File: clojure/world.clj
(ns world
  (:require [cell :as c]
            [clojure.string :as string]))

(defrecord World [tick width height cells])

(defn location-occupied-exception [x y]
  (ex-info (str "LocationOccupied(" x "-" y ")") {:x x :y y}))

(def directions
  [[-1 1]  [0 1]  [1 1]  ; above
   [-1 0]         [1 0]  ; sides
   [-1 -1] [0 -1] [1 -1] ; below
   ])

(declare make-key cell-at populate-cells add-cell prepopulate-neighbours)

(defn new-world [width height]
  (let [world (->World 0 width height {})]
    (-> world
        populate-cells
        prepopulate-neighbours)))

(defn tick [world]
  (let [cells (:cells world)
        updated-cells
        (persistent!
         (reduce-kv
          (fn [acc key cell]
            (let [alive-neighbours (c/alive-neighbours cell world)]
              (cond
                (and (not (:alive cell))
                     (= alive-neighbours 3))
                (assoc! acc key (assoc cell :alive true))

                (or (< alive-neighbours 2)
                    (> alive-neighbours 3))
                (assoc! acc key (assoc cell :alive false))

                :else
                (assoc! acc key cell))))
          (transient {})
          cells))]
    (-> world
        (assoc :cells updated-cells)
        (update :tick inc))))

(defn render [world]
  ;; The following is slower
  ;; (loop [y 0
  ;;        rendering ""]
  ;;   (if (>= y (:height world))
  ;;     rendering
  ;;     (recur (inc y)
  ;;            (loop [x 0
  ;;                   line rendering]
  ;;              (if (>= x (:width world))
  ;;                (str line "\n")
  ;;                (let [cell (cell-at world x y)]
  ;;                  (recur (inc x)
  ;;                         (if cell
  ;;                           (str line (c/to-char cell))
  ;;                           line))))))))

  ;; The following is slower
  ;; (let [chars (loop [y 0
  ;;                    result []]
  ;;               (if (>= y (:height world))
  ;;                 result
  ;;                 (recur (inc y)
  ;;                        (loop [x 0
  ;;                               line result]
  ;;                          (if (>= x (:width world))
  ;;                            (conj line \newline)
  ;;                            (let [cell (cell-at world x y)]
  ;;                              (recur (inc x)
  ;;                                     (if cell
  ;;                                       (conj line (c/to-char cell))
  ;;                                       line))))))))]
  ;;   (apply str chars))

  ;; The following is the fastest
  (let [render-size (+ (* (:width world) (:height world)) (:height world))
        rendering (StringBuilder. (int render-size))]
    (doseq [y (range (:height world))]
      (doseq [x (range (:width world))]
        (let [cell (cell-at world x y)]
          (when cell
            (.append rendering (c/to-char cell)))))
      (.append rendering \newline))
    (.toString rendering)))

(defn- make-key [x y]
  ;; The following is slower
  ;; (format "%d-%d" x y)

  ;; The following is slower
  ;; (str x "-" y)

  ;; The following is the fastest
  (string/join "-" [x y]))

(defn- cell-at [world x y]
  (let [key (make-key x y)]
    (get (:cells world) key)))

(defn- populate-cells [world]
  (reduce (fn [world y]
            (reduce (fn [world x]
                      (let [alive (<= (rand) 0.2)
                            [world _success] (add-cell world x y alive)]
                        world))
                    world
                    (range (:width world))))
          world
          (range (:height world))))

(defn- add-cell
  ([world x y]
   (add-cell world x y false))
  ([world x y alive]
   (let [existing (cell-at world x y)]
     (if existing
       (throw (location-occupied-exception x y))
       (let [key (make-key x y)
             cell (c/new-cell x y alive)]
         [(assoc-in world [:cells key] cell) true])))))

(defn- prepopulate-neighbours [world]
  (reduce-kv
   (fn [world key cell]
     (let [neighbours
           (vec (keep (fn [[rel-x rel-y]]
                        (let [nx (+ (:x cell) rel-x)
                              ny (+ (:y cell) rel-y)]
                          (when-not (or (< nx 0) (< ny 0)) ; Out of bounds
                            (when-not (or (>= nx (:width world)) (>= ny (:height world))) ; Out of bounds
                              (let [neighbour-key (make-key nx ny)]
                                (when (get (:cells world) neighbour-key)
                                  neighbour-key))))))
                      directions))]
       (assoc-in world [:cells key :neighbours] neighbours)))
   world
   (:cells world)))
